File: CK3ToEU4a/Source/CK3World/Dynasties/House.cpp
#include "House.h"
#include "CommonRegexes.h"
#include "Log.h"
#include "ParserHelpers.h"

CK3::House::House(std::istream& theStream, long long housID): houseID(housID)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void CK3::House::registerKeys()
{
	registerKeyword("key", [this](std::istream& theStream) {
		key = commonItems::getString(theStream);
	});
	registerKeyword("name", [this](const std::string& unused, std::istream& theStream) {
		name = commonItems::singleString(theStream).getString();
	});
	registerKeyword("localized_name", [this](const std::string& unused, std::istream& theStream) {
		localizedName = commonItems::singleString(theStream).getString();
	});
	registerKeyword("prefix", [this](const std::string& unused, std::istream& theStream) {
		prefix = commonItems::singleString(theStream).getString();
	});
	registerKeyword("dynasty", [this](const std::string& unused, std::istream& theStream) {
		dynasty = std::make_pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerKeyword("head_of_house", [this](const std::string& unused, std::istream& theStream) {
		houseHead = std::make_pair(commonItems::singleLlong(theStream).getLlong(), nullptr);
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: CK3ToEU4/Source/CK3/Dynasties/House.cs

using System.Collections.Generic;
using CK3;
using commonItems;

namespace CK3ToEU4.CK3.Dynasties;

class House
{

	public House()
	{
	}

	public House(BufferedReader reader, long houseID)
	{
		ID = houseID;
		
		var parser = new Parser();
		registerKeys(parser);
		parser.ParseStream(reader);
	}
	public string getKey() { return key; }
	public string getName() { return name; }
	public string getLocalizedName() { return localizedName; }
	public string getPrefix() { return prefix; }
	public KeyValuePair<long, Dynasty?> getDynasty() { return dynasty; }
	public long ID { get; private set; } = 0;
	public KeyValuePair<long, Character?>? getHouseHead() { return houseHead; }

	public void loadDynasty(KeyValuePair<long, Dynasty?> theDynasty) { dynasty = theDynasty; }
	public void loadHouseHead(KeyValuePair<long, Character?> theHead) { houseHead = theHead; }

	public void resetHouseHead()
	{
		houseHead = null;
	}

	public void setName(string theName) { name = theName; }
	public void setPrefix(string thePrefix) { prefix = thePrefix; }

  
	private void registerKeys(Parser parser)
	{
		parser.RegisterKeyword("key", reader => {
			key = reader.GetString();
		});
		parser.RegisterKeyword("name", reader => {
			name = reader.GetString();
		});
		parser.RegisterKeyword("localized_name", reader => {
			localizedName = reader.GetString();
		});
		parser.RegisterKeyword("prefix", reader => {
			prefix = reader.GetString();
		});
		parser.RegisterKeyword("dynasty", reader => {
			dynasty = new(reader.GetLong(), null);
		});
		parser.RegisterKeyword("head_of_house", reader => {
			houseHead = new(reader.GetLong(), null);
		});
		parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
	}
	
	private string key;
	private string name;
	private string localizedName;
	private KeyValuePair<long, Dynasty?> dynasty;
	private string prefix;
	private KeyValuePair<long, Character?>? houseHead; // houses can have missing heads or dead people...
};